File: data-structures/linear/stack/Stack.go
package main

import (
	"errors"
)

type StackNode struct {
	data int
	next *StackNode
}

type Stack struct {
	head *StackNode
	size int
}

func (s *Stack) Push(data int) {
	newNode := &StackNode{
		data: data,
	}
	newNode.next = s.head
	s.head = newNode
	s.size++
}

func (s *Stack) Pop() (int, error) {
	if s.head == nil {
		return 0, errors.New("stack is empty")
	}
	val := s.head.data
	s.head = s.head.next
	s.size--
	return val, nil
}

func (s *Stack) Top() (int, error) {
	if s.head == nil {
		return 0, errors.New("stack is empty")
	}
	return s.head.data, nil
}

func (s *Stack) Size() int {
	return s.size
}

func (s *Stack) IsEmpty() bool {
	return s.size == 0
}

File: data-structures/linear/stack/Stack.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    def __init__(self):
        self.head = None
        self._size = 0

    def push(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node
        self._size += 1

    def pop(self):
        if self.head is None:
            print("Stack Underflow")
            return
        self.head = self.head.next
        self._size -= 1

    def top(self):
        if self.head is None:
            print("Stack Underflow")
            return
        print(self.head.data)

    def is_empty(self):
        return self.head is None

    def size(self):
        return self._size
